File: python/chancery_reference/binary_codec.py
from __future__ import annotations

import base64
from collections.abc import Mapping, Sequence
from typing import cast

from .base58_codec import decode_base58, encode_base58
from .schema import SchemaRecord, TypeSchema, defined_type_schema

StructValues = Mapping[str, object]


def _integer_from_value(value: object, label: str) -> int:
    if isinstance(value, bool):
        raise ValueError(f"{label} must be an integer")
    if isinstance(value, int):
        return value
    if isinstance(value, str) and value.lstrip("-").isdigit():
        return int(value, 10)
    raise ValueError(f"{label} must be an integer or decimal integer string")


def _encode_unsigned_integer(value: object, byte_length: int, label: str) -> bytes:
    integer_value = _integer_from_value(value, label)
    maximum_value = 1 << (byte_length * 8)
    if integer_value < 0 or integer_value >= maximum_value:
        raise ValueError(f"{label} is outside the unsigned {byte_length * 8}-bit range")
    return integer_value.to_bytes(byte_length, "little", signed=False)


def _encode_signed_integer(value: object, byte_length: int, label: str) -> bytes:
    integer_value = _integer_from_value(value, label)
    minimum_value = -(1 << (byte_length * 8 - 1))
    maximum_value = (1 << (byte_length * 8 - 1)) - 1
    if integer_value < minimum_value or integer_value > maximum_value:
        raise ValueError(f"{label} is outside the signed {byte_length * 8}-bit range")
    return integer_value.to_bytes(byte_length, "little", signed=True)


def decode_base64(encoded: str) -> bytes:
    return base64.b64decode(encoded, validate=True)


def encode_base64(data: bytes | bytearray | memoryview) -> str:
    return base64.b64encode(bytes(data)).decode("ascii")


def bytes_from_value(value: object, label: str) -> bytes:
    if isinstance(value, (bytes, bytearray, memoryview)):
        return bytes(value)
    if isinstance(value, Sequence) and not isinstance(value, str):
        byte_values = bytearray()
        for index, byte_value in enumerate(value):
            if isinstance(byte_value, bool) or not isinstance(byte_value, int) or byte_value < 0 or byte_value > 255:
                raise ValueError(f"{label}[{index}] must be an unsigned byte")
            byte_values.append(byte_value)
        return bytes(byte_values)
    if isinstance(value, str):
        if value.startswith("0x"):
            try:
                return bytes.fromhex(value[2:])
            except ValueError as error:
                raise ValueError(f"{label} contains invalid hexadecimal") from error
        if value.startswith("base64:"):
            return decode_base64(value[7:])
    raise ValueError(
        f"{label} must be bytes, an unsigned-byte sequence, 0x-prefixed hexadecimal, or base64-prefixed data"
    )


def _type_kind(type_schema: SchemaRecord) -> str:
    kind = type_schema.get("kind")
    if not isinstance(kind, str):
        raise ValueError("Type schema kind must be a string")
    return kind


def _type_item(type_schema: SchemaRecord) -> TypeSchema:
    item = type_schema.get("item")
    if not isinstance(item, (str, dict)):
        raise ValueError("Type schema item is missing")
    return cast(TypeSchema, item)


def _field_schemas(struct_schema: SchemaRecord) -> list[SchemaRecord]:
    fields_value = struct_schema.get("fields")
    if not isinstance(fields_value, list):
        raise ValueError("Struct schema fields must be an array")
    fields: list[SchemaRecord] = []
    for field_value in fields_value:
        if not isinstance(field_value, dict):
            raise ValueError("Struct field schema must be an object")
        fields.append(cast(SchemaRecord, field_value))
    return fields


def _encode_primitive(type_name: str, value: object, label: str) -> bytes:
    if type_name == "u8":
        return _encode_unsigned_integer(value, 1, label)
    if type_name == "u16":
        return _encode_unsigned_integer(value, 2, label)
    if type_name == "u32":
        return _encode_unsigned_integer(value, 4, label)
    if type_name == "u64":
        return _encode_unsigned_integer(value, 8, label)
    if type_name == "u128":
        return _encode_unsigned_integer(value, 16, label)
    if type_name == "i64":
        return _encode_signed_integer(value, 8, label)
    if type_name == "i128":
        return _encode_signed_integer(value, 16, label)
    if type_name == "bool":
        if not isinstance(value, bool):
            raise ValueError(f"{label} must be a boolean")
        return b"\x01" if value else b"\x00"
    if type_name == "pubkey":
        public_key = decode_base58(value) if isinstance(value, str) else bytes_from_value(value, label)
        if len(public_key) != 32:
            raise ValueError(f"{label} public key must contain 32 bytes")
        return public_key
    raise ValueError(f"Unsupported primitive type: {type_name}")


def encode_type(type_schema: TypeSchema, value: object, label: str = "value") -> bytes:
    if isinstance(type_schema, str):
        return _encode_primitive(type_schema, value, label)

    kind = _type_kind(type_schema)
    if kind == "option":
        if value is None:
            return b"\x00"
        return b"\x01" + encode_type(_type_item(type_schema), value, label)
    if kind == "array":
        length_value = type_schema.get("length")
        if not isinstance(length_value, int):
            raise ValueError("Array type length must be an integer")
        item_schema = _type_item(type_schema)
        if item_schema == "u8":
            data = bytes_from_value(value, label)
            if len(data) != length_value:
                raise ValueError(f"{label} must contain exactly {length_value} bytes")
            return data
        if not isinstance(value, Sequence) or isinstance(value, str):
            raise ValueError(f"{label} must be an array")
        if len(value) != length_value:
            raise ValueError(f"{label} must contain exactly {length_value} items")
        return b"".join(
            encode_type(item_schema, item_value, f"{label}[{index}]")
            for index, item_value in enumerate(value)
        )
    if kind == "vector":
        item_schema = _type_item(type_schema)
        if item_schema == "u8":
            data = bytes_from_value(value, label)
            return _encode_unsigned_integer(len(data), 4, f"{label}.length") + data
        if not isinstance(value, Sequence) or isinstance(value, str):
            raise ValueError(f"{label} must be an array")
        return _encode_unsigned_integer(len(value), 4, f"{label}.length") + b"".join(
            encode_type(item_schema, item_value, f"{label}[{index}]")
            for index, item_value in enumerate(value)
        )
    if kind == "defined":
        type_name = type_schema.get("name")
        if not isinstance(type_name, str):
            raise ValueError("Defined type name must be a string")
        if not isinstance(value, Mapping):
            raise ValueError(f"{label} must be an object")
        return encode_struct(defined_type_schema(type_name), cast(StructValues, value), label)
    raise ValueError(f"Unsupported type schema kind: {kind}")


def zero_value_for_type(type_schema: TypeSchema) -> object:
    if isinstance(type_schema, str):
        if type_schema == "bool":
            return False
        if type_schema == "pubkey":
            return bytes(32)
        return 0
    kind = _type_kind(type_schema)
    if kind == "option":
        return None
    if kind == "vector":
        return b"" if _type_item(type_schema) == "u8" else []
    if kind == "array":
        length_value = type_schema.get("length")
        if not isinstance(length_value, int):
            raise ValueError("Array type length must be an integer")
        item_schema = _type_item(type_schema)
        if item_schema == "u8":
            return bytes(length_value)
        return [zero_value_for_type(item_schema) for _ in range(length_value)]
    if kind == "defined":
        type_name = type_schema.get("name")
        if not isinstance(type_name, str):
            raise ValueError("Defined type name must be a string")
        values: dict[str, object] = {}
        for field_schema in _field_schemas(defined_type_schema(type_name)):
            field_name = field_schema.get("name")
            field_type = field_schema.get("type")
            if not isinstance(field_name, str) or not isinstance(field_type, (str, dict)):
                raise ValueError("Defined type field schema is invalid")
            values[field_name] = zero_value_for_type(cast(TypeSchema, field_type))
        return values
    raise ValueError(f"Unsupported type schema kind: {kind}")


def encode_struct(
    struct_schema: SchemaRecord,
    values: StructValues,
    label: str = "struct",
    default_missing_padding: bool = False,
) -> bytes:
    encoded_fields: list[bytes] = []
    for field_schema in _field_schemas(struct_schema):
        field_name = field_schema.get("name")
        field_type = field_schema.get("type")
        if not isinstance(field_name, str) or not isinstance(field_type, (str, dict)):
            raise ValueError("Struct field schema is invalid")
        typed_field = cast(TypeSchema, field_type)
        field_value = values.get(field_name)
        if field_name not in values and default_missing_padding and (
            field_name.startswith("_pad") or field_name.startswith("_reserved")
        ):
            field_value = zero_value_for_type(typed_field)
        if field_name not in values and not (
            isinstance(typed_field, dict) and typed_field.get("kind") == "option"
        ) and field_value is None:
            raise ValueError(f"{label}.{field_name} is required")
        encoded_fields.append(encode_type(typed_field, field_value, f"{label}.{field_name}"))
    return b"".join(encoded_fields)


class BinaryReader:
    def __init__(self, data: bytes | bytearray | memoryview, offset: int = 0) -> None:
        self._data = bytes(data)
        if offset < 0 or offset > len(self._data):
            raise ValueError("BinaryReader offset is outside the input range")
        self._offset = offset

    @property
    def offset(self) -> int:
        return self._offset

    @property
    def remaining(self) -> int:
        return len(self._data) - self._offset

    def read_bytes(self, byte_length: int, label: str) -> bytes:
        if byte_length < 0 or self._offset + byte_length > len(self._data):
            raise ValueError(
                f"{label} requires {byte_length} bytes with only {self.remaining} remaining"
            )
        data = self._data[self._offset : self._offset + byte_length]
        self._offset += byte_length
        return data


def _decode_primitive(type_name: str, reader: BinaryReader, label: str) -> object:
    if type_name == "u8":
        return int.from_bytes(reader.read_bytes(1, label), "little", signed=False)
    if type_name == "u16":
        return int.from_bytes(reader.read_bytes(2, label), "little", signed=False)
    if type_name == "u32":
        return int.from_bytes(reader.read_bytes(4, label), "little", signed=False)
    if type_name == "u64":
        return int.from_bytes(reader.read_bytes(8, label), "little", signed=False)
    if type_name == "u128":
        return int.from_bytes(reader.read_bytes(16, label), "little", signed=False)
    if type_name == "i64":
        return int.from_bytes(reader.read_bytes(8, label), "little", signed=True)
    if type_name == "i128":
        return int.from_bytes(reader.read_bytes(16, label), "little", signed=True)
    if type_name == "bool":
        value = int.from_bytes(reader.read_bytes(1, label), "little", signed=False)
        if value not in (0, 1):
            raise ValueError(f"{label} contains an invalid boolean tag: {value}")
        return value == 1
    if type_name == "pubkey":
        return encode_base58(reader.read_bytes(32, label))
    raise ValueError(f"Unsupported primitive type: {type_name}")


def decode_type(type_schema: TypeSchema, reader: BinaryReader, label: str = "value") -> object:
    if isinstance(type_schema, str):
        return _decode_primitive(type_schema, reader, label)

    kind = _type_kind(type_schema)
    if kind == "option":
        tag = int.from_bytes(reader.read_bytes(1, f"{label}.tag"), "little")
        if tag == 0:
            return None
        if tag != 1:
            raise ValueError(f"{label} contains an invalid option tag: {tag}")
        return decode_type(_type_item(type_schema), reader, label)
    if kind == "array":
        length_value = type_schema.get("length")
        if not isinstance(length_value, int):
            raise ValueError("Array type length must be an integer")
        item_schema = _type_item(type_schema)
        if item_schema == "u8":
            return reader.read_bytes(length_value, label)
        return [
            decode_type(item_schema, reader, f"{label}[{index}]")
            for index in range(length_value)
        ]
    if kind == "vector":
        length_value = int.from_bytes(reader.read_bytes(4, f"{label}.length"), "little")
        item_schema = _type_item(type_schema)
        if item_schema == "u8":
            return reader.read_bytes(length_value, label)
        return [
            decode_type(item_schema, reader, f"{label}[{index}]")
            for index in range(length_value)
        ]
    if kind == "defined":
        type_name = type_schema.get("name")
        if not isinstance(type_name, str):
            raise ValueError("Defined type name must be a string")
        return decode_struct(defined_type_schema(type_name), reader, label)
    raise ValueError(f"Unsupported type schema kind: {kind}")


def decode_struct(
    struct_schema: SchemaRecord,
    reader: BinaryReader,
    label: str = "struct",
) -> dict[str, object]:
    values: dict[str, object] = {}
    for field_schema in _field_schemas(struct_schema):
        field_name = field_schema.get("name")
        field_type = field_schema.get("type")
        if not isinstance(field_name, str) or not isinstance(field_type, (str, dict)):
            raise ValueError("Struct field schema is invalid")
        values[field_name] = decode_type(
            cast(TypeSchema, field_type), reader, f"{label}.{field_name}"
        )
    return values


def field_schema_by_name(fields_value: object, field_name: str) -> SchemaRecord:
    if not isinstance(fields_value, list):
        raise ValueError("Fields schema must be an array")
    for field_value in fields_value:
        if isinstance(field_value, dict) and field_value.get("name") == field_name:
            return cast(SchemaRecord, field_value)
    raise ValueError(f"Unknown field: {field_name}")


def to_json_compatible(value: object) -> object:
    if isinstance(value, bytes):
        return f"0x{value.hex()}"
    if isinstance(value, dict):
        return {str(key): to_json_compatible(item) for key, item in value.items()}
    if isinstance(value, list):
        return [to_json_compatible(item) for item in value]
    if isinstance(value, tuple):
        return [to_json_compatible(item) for item in value]
    return value

File: typescript/src/BinaryCodec.ts
import { decodeBase58, encodeBase58 } from "./Base58Codec.js";
import type { FieldSchema, StructSchema, TypeSchema } from "./ChancerySchema.js";
import { getDefinedTypeSchema } from "./ChancerySchema.js";

export type StructValues = Readonly<Record<string, unknown>>;

function concatenateBytes(parts: readonly Uint8Array[]): Uint8Array {
    let totalLength = 0;
    for (let index = 0, length = parts.length; index < length; index++) {
        totalLength += parts[index]?.length ?? 0;
    }
    const bytes = new Uint8Array(totalLength);
    let offset = 0;
    for (let index = 0, length = parts.length; index < length; index++) {
        const part = parts[index];
        if (part === undefined) {
            continue;
        }
        bytes.set(part, offset);
        offset += part.length;
    }
    return bytes;
}

function integerFromValue(value: unknown, label: string): bigint {
    if (typeof value === "bigint") {
        return value;
    }
    if (typeof value === "number" && Number.isSafeInteger(value)) {
        return BigInt(value);
    }
    if (typeof value === "string" && /^-?[0-9]+$/.test(value)) {
        return BigInt(value);
    }
    throw new Error(`${label} must be a bigint, safe integer, or decimal integer string`);
}

function encodeUnsignedInteger(value: unknown, byteLength: number, label: string): Uint8Array {
    let integerValue = integerFromValue(value, label);
    const maximumValue = 1n << BigInt(byteLength * 8);
    if (integerValue < 0n || integerValue >= maximumValue) {
        throw new Error(`${label} is outside the unsigned ${byteLength * 8}-bit range`);
    }
    const bytes = new Uint8Array(byteLength);
    for (let index = 0; index < byteLength; index++) {
        bytes[index] = Number(integerValue & 0xffn);
        integerValue >>= 8n;
    }
    return bytes;
}

function encodeSignedInteger(value: unknown, byteLength: number, label: string): Uint8Array {
    let integerValue = integerFromValue(value, label);
    const bitLength = BigInt(byteLength * 8);
    const minimumValue = -(1n << (bitLength - 1n));
    const maximumValue = (1n << (bitLength - 1n)) - 1n;
    if (integerValue < minimumValue || integerValue > maximumValue) {
        throw new Error(`${label} is outside the signed ${byteLength * 8}-bit range`);
    }
    if (integerValue < 0n) {
        integerValue += 1n << bitLength;
    }
    return encodeUnsignedInteger(integerValue, byteLength, label);
}

function decodeHex(hexadecimal: string): Uint8Array {
    const body = hexadecimal.startsWith("0x") ? hexadecimal.slice(2) : hexadecimal;
    if (body.length % 2 !== 0 || !/^[0-9a-fA-F]*$/.test(body)) {
        throw new Error("Hexadecimal byte strings require an even number of hexadecimal digits");
    }
    const bytes = new Uint8Array(body.length / 2);
    for (let index = 0, length = bytes.length; index < length; index++) {
        bytes[index] = Number.parseInt(body.slice(index * 2, index * 2 + 2), 16);
    }
    return bytes;
}

export function decodeBase64(encoded: string): Uint8Array {
    const binary = atob(encoded);
    const bytes = new Uint8Array(binary.length);
    for (let index = 0, length = binary.length; index < length; index++) {
        bytes[index] = binary.charCodeAt(index);
    }
    return bytes;
}

export function encodeBase64(bytes: Uint8Array): string {
    const chunkSize = 0x8000;
    let binary = "";
    for (let offset = 0; offset < bytes.length; offset += chunkSize) {
        const end = Math.min(offset + chunkSize, bytes.length);
        for (let index = offset; index < end; index++) {
            binary += String.fromCharCode(bytes[index] ?? 0);
        }
    }
    return btoa(binary);
}

export function bytesFromValue(value: unknown, label: string): Uint8Array {
    if (value instanceof Uint8Array) {
        return new Uint8Array(value);
    }
    if (Array.isArray(value)) {
        const bytes = new Uint8Array(value.length);
        for (let index = 0, length = value.length; index < length; index++) {
            const byteValue = value[index];
            if (typeof byteValue !== "number" || !Number.isInteger(byteValue) || byteValue < 0 || byteValue > 255) {
                throw new Error(`${label}[${index}] must be an unsigned byte`);
            }
            bytes[index] = byteValue;
        }
        return bytes;
    }
    if (typeof value === "string") {
        if (value.startsWith("0x")) {
            return decodeHex(value);
        }
        if (value.startsWith("base64:")) {
            return decodeBase64(value.slice(7));
        }
    }
    throw new Error(
        `${label} must be Uint8Array, an unsigned-byte array, ` +
        "0x-prefixed hexadecimal, or base64-prefixed data",
    );
}

function recordFromValue(value: unknown, label: string): Readonly<Record<string, unknown>> {
    if (typeof value !== "object" || value === null || Array.isArray(value) || value instanceof Uint8Array) {
        throw new Error(`${label} must be an object`);
    }
    return value as Readonly<Record<string, unknown>>;
}

function arrayFromValue(value: unknown, label: string): readonly unknown[] {
    if (!Array.isArray(value)) {
        throw new Error(`${label} must be an array`);
    }
    return value;
}

function encodePrimitive(typeName: string, value: unknown, label: string): Uint8Array {
    switch (typeName) {
        case "u8":
            return encodeUnsignedInteger(value, 1, label);
        case "u16":
            return encodeUnsignedInteger(value, 2, label);
        case "u32":
            return encodeUnsignedInteger(value, 4, label);
        case "u64":
            return encodeUnsignedInteger(value, 8, label);
        case "u128":
            return encodeUnsignedInteger(value, 16, label);
        case "i64":
            return encodeSignedInteger(value, 8, label);
        case "i128":
            return encodeSignedInteger(value, 16, label);
        case "bool":
            if (typeof value !== "boolean") {
                throw new Error(`${label} must be a boolean`);
            }
            return new Uint8Array([value ? 1 : 0]);
        case "pubkey": {
            const publicKeyBytes = typeof value === "string" ? decodeBase58(value) : bytesFromValue(value, label);
            if (publicKeyBytes.length !== 32) {
                throw new Error(`${label} public key must contain 32 bytes`);
            }
            return publicKeyBytes;
        }
        default:
            throw new Error(`Unsupported primitive type: ${typeName}`);
    }
}

export function encodeType(typeSchema: TypeSchema, value: unknown, label = "value"): Uint8Array {
    if (typeof typeSchema === "string") {
        return encodePrimitive(typeSchema, value, label);
    }
    if (typeSchema.kind === "option") {
        if (value === null || value === undefined) {
            return new Uint8Array([0]);
        }
        return concatenateBytes([new Uint8Array([1]), encodeType(typeSchema.item, value, label)]);
    }
    if (typeSchema.kind === "array") {
        if (typeSchema.item === "u8") {
            const bytes = bytesFromValue(value, label);
            if (bytes.length !== typeSchema.length) {
                throw new Error(`${label} must contain exactly ${typeSchema.length} bytes`);
            }
            return bytes;
        }
        const values = arrayFromValue(value, label);
        if (values.length !== typeSchema.length) {
            throw new Error(`${label} must contain exactly ${typeSchema.length} items`);
        }
        const parts: Uint8Array[] = [];
        for (let index = 0; index < typeSchema.length; index++) {
            parts.push(encodeType(typeSchema.item, values[index], `${label}[${index}]`));
        }
        return concatenateBytes(parts);
    }
    if (typeSchema.kind === "vector") {
        if (typeSchema.item === "u8") {
            const bytes = bytesFromValue(value, label);
            return concatenateBytes([encodeUnsignedInteger(bytes.length, 4, `${label}.length`), bytes]);
        }
        const values = arrayFromValue(value, label);
        const parts: Uint8Array[] = [encodeUnsignedInteger(values.length, 4, `${label}.length`)];
        for (let index = 0, length = values.length; index < length; index++) {
            parts.push(encodeType(typeSchema.item, values[index], `${label}[${index}]`));
        }
        return concatenateBytes(parts);
    }
    if (typeSchema.kind === "defined") {
        return encodeStruct(getDefinedTypeSchema(typeSchema.name), recordFromValue(value, label), label);
    }
    throw new Error("Unsupported type schema");
}

export function zeroValueForType(typeSchema: TypeSchema): unknown {
    if (typeof typeSchema === "string") {
        switch (typeSchema) {
            case "bool":
                return false;
            case "pubkey":
                return new Uint8Array(32);
            default:
                return 0;
        }
    }
    if (typeSchema.kind === "option") {
        return null;
    }
    if (typeSchema.kind === "vector") {
        return typeSchema.item === "u8" ? new Uint8Array(0) : [];
    }
    if (typeSchema.kind === "array") {
        if (typeSchema.item === "u8") {
            return new Uint8Array(typeSchema.length);
        }
        const values: unknown[] = [];
        for (let index = 0; index < typeSchema.length; index++) {
            values.push(zeroValueForType(typeSchema.item));
        }
        return values;
    }
    if (typeSchema.kind === "defined") {
        const values: Record<string, unknown> = {};
        const structSchema = getDefinedTypeSchema(typeSchema.name);
        for (let index = 0, length = structSchema.fields.length; index < length; index++) {
            const field = structSchema.fields[index];
            if (field !== undefined) {
                values[field.name] = zeroValueForType(field.type);
            }
        }
        return values;
    }
    throw new Error("Unsupported type schema");
}

export function encodeStruct(
    structSchema: StructSchema,
    values: StructValues,
    label = "struct",
    defaultMissingPadding = false,
): Uint8Array {
    const parts: Uint8Array[] = [];
    for (let index = 0, length = structSchema.fields.length; index < length; index++) {
        const field = structSchema.fields[index];
        if (field === undefined) {
            continue;
        }
        let fieldValue = values[field.name];
        if (
            fieldValue === undefined &&
            defaultMissingPadding &&
            (field.name.startsWith("_pad") || field.name.startsWith("_reserved"))
        ) {
            fieldValue = zeroValueForType(field.type);
        }
        if (fieldValue === undefined && !(typeof field.type !== "string" && field.type.kind === "option")) {
            throw new Error(`${label}.${field.name} is required`);
        }
        parts.push(encodeType(field.type, fieldValue, `${label}.${field.name}`));
    }
    return concatenateBytes(parts);
}

function decodeUnsignedInteger(bytes: Uint8Array): bigint {
    let value = 0n;
    for (let index = bytes.length - 1; index >= 0; index--) {
        value = (value << 8n) | BigInt(bytes[index] ?? 0);
    }
    return value;
}

function decodeSignedInteger(bytes: Uint8Array): bigint {
    const unsignedValue = decodeUnsignedInteger(bytes);
    const bitLength = BigInt(bytes.length * 8);
    const signBit = 1n << (bitLength - 1n);
    return (unsignedValue & signBit) === 0n ? unsignedValue : unsignedValue - (1n << bitLength);
}

export class BinaryReader {
    readonly #bytes: Uint8Array;
    #offset = 0;

    constructor(bytes: Uint8Array, offset = 0) {
        if (!Number.isInteger(offset) || offset < 0 || offset > bytes.length) {
            throw new Error("BinaryReader offset is outside the input range");
        }
        this.#bytes = bytes;
        this.#offset = offset;
    }

    get offset(): number {
        return this.#offset;
    }

    get remaining(): number {
        return this.#bytes.length - this.#offset;
    }

    readBytes(byteLength: number, label: string): Uint8Array {
        if (!Number.isInteger(byteLength) || byteLength < 0 || this.#offset + byteLength > this.#bytes.length) {
            throw new Error(`${label} requires ${byteLength} bytes with only ${this.remaining} remaining`);
        }
        const bytes = this.#bytes.slice(this.#offset, this.#offset + byteLength);
        this.#offset += byteLength;
        return bytes;
    }
}

function decodePrimitive(typeName: string, reader: BinaryReader, label: string): unknown {
    switch (typeName) {
        case "u8":
            return Number(decodeUnsignedInteger(reader.readBytes(1, label)));
        case "u16":
            return Number(decodeUnsignedInteger(reader.readBytes(2, label)));
        case "u32":
            return Number(decodeUnsignedInteger(reader.readBytes(4, label)));
        case "u64":
            return decodeUnsignedInteger(reader.readBytes(8, label));
        case "u128":
            return decodeUnsignedInteger(reader.readBytes(16, label));
        case "i64":
            return decodeSignedInteger(reader.readBytes(8, label));
        case "i128":
            return decodeSignedInteger(reader.readBytes(16, label));
        case "bool": {
            const value = Number(decodeUnsignedInteger(reader.readBytes(1, label)));
            if (value !== 0 && value !== 1) {
                throw new Error(`${label} contains an invalid boolean tag: ${value}`);
            }
            return value === 1;
        }
        case "pubkey":
            return encodeBase58(reader.readBytes(32, label));
        default:
            throw new Error(`Unsupported primitive type: ${typeName}`);
    }
}

export function decodeType(typeSchema: TypeSchema, reader: BinaryReader, label = "value"): unknown {
    if (typeof typeSchema === "string") {
        return decodePrimitive(typeSchema, reader, label);
    }
    if (typeSchema.kind === "option") {
        const tag = Number(decodeUnsignedInteger(reader.readBytes(1, `${label}.tag`)));
        if (tag === 0) {
            return null;
        }
        if (tag !== 1) {
            throw new Error(`${label} contains an invalid option tag: ${tag}`);
        }
        return decodeType(typeSchema.item, reader, label);
    }
    if (typeSchema.kind === "array") {
        if (typeSchema.item === "u8") {
            return reader.readBytes(typeSchema.length, label);
        }
        const values: unknown[] = [];
        for (let index = 0; index < typeSchema.length; index++) {
            values.push(decodeType(typeSchema.item, reader, `${label}[${index}]`));
        }
        return values;
    }
    if (typeSchema.kind === "vector") {
        const length = Number(decodeUnsignedInteger(reader.readBytes(4, `${label}.length`)));
        if (typeSchema.item === "u8") {
            return reader.readBytes(length, label);
        }
        const values: unknown[] = [];
        for (let index = 0; index < length; index++) {
            values.push(decodeType(typeSchema.item, reader, `${label}[${index}]`));
        }
        return values;
    }
    if (typeSchema.kind === "defined") {
        return decodeStruct(getDefinedTypeSchema(typeSchema.name), reader, label);
    }
    throw new Error("Unsupported type schema");
}

export function decodeStruct(
    structSchema: StructSchema,
    reader: BinaryReader,
    label = "struct",
): Record<string, unknown> {
    const values: Record<string, unknown> = {};
    for (let index = 0, length = structSchema.fields.length; index < length; index++) {
        const field = structSchema.fields[index];
        if (field !== undefined) {
            values[field.name] = decodeType(field.type, reader, `${label}.${field.name}`);
        }
    }
    return values;
}

export function fieldSchemaByName(fields: readonly FieldSchema[], fieldName: string): FieldSchema {
    for (let index = 0, length = fields.length; index < length; index++) {
        const field = fields[index];
        if (field?.name === fieldName) {
            return field;
        }
    }
    throw new Error(`Unknown field: ${fieldName}`);
}

export function bytesToHex(bytes: Uint8Array): string {
    let hexadecimal = "";
    for (let index = 0, length = bytes.length; index < length; index++) {
        hexadecimal += (bytes[index] ?? 0).toString(16).padStart(2, "0");
    }
    return hexadecimal;
}

export function chanceryJsonReplacer(_key: string, value: unknown): unknown {
    if (typeof value === "bigint") {
        return value.toString();
    }
    if (value instanceof Uint8Array) {
        return `0x${bytesToHex(value)}`;
    }
    return value;
}
